### PR TITLE
feat (DOCS-CODE-035): [authN] [sign-in] canonical X-Plat identity sample for user sign-in using in .NET 6 MAUI

### PR DESCRIPTION
**Pull request by ferantivero**
closes: #513513, #1670076, #543974

## WHY

we wanted to create our canonical sample for user sign-in using a MAUI .NET. This sample is meant to be used from Microsoft Identity platform tutorials.

## WHAT Changed?

- add canonical maui app
- add user sign-in/out using `Microsoft.Identity.Client` 
- add readme following recent guidelines

## HOW to test?

Follow the instructions in the `README.md` file.

### CODE REVIEW

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (ckittel) [inline src/sign-in-xplat/MainPage.xaml.cs:41]: I'm going to make the suggestion before Marsh does (and this is based off of the recent feedback on the python webapp work), instead of hiding this in a helper method, please add it all inline here.  Specifically to help the user see (without jumping context) what code they are responsible for.  I had to do the same with the Python app code, creating very "fat" (controller) methods just to be very clear about the "flow" of things.  What you have here is closer to what we would normally write day-to-day, but not what we want for the tutorial series, where possible.

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (ckittel) [inline src/sign-in-xplat/MainPage.xaml.cs:125]: `PuiblicClientApp`'s lifecycle, I would image, should be scoped closer to the lifecycle of the app.  In the way that you're presenting it here, you'll always need to sign in and will never leverage the token cache from what I can tell.  Maybe you move the `PublicClientApp` lifecycle out to app scope, then a call like this line would be cheaper as it would leverage the token cache by default and only fall back to interactive when necessary (instead of always).

The other issue with `new`ing up a `PublicClientApp` is that the HTTP request to the OIDC metadata endpoint would happen each time, when in reality that http client response would also be cached.

Maybe I'm wrong with all of this and it's somehow caching it all, but I'm not quite following how that would be.

**Review by ckittel — CHANGES_REQUESTED** [review]
Peer review engineering feedback below. I mostly ignored the README file as it seems CD will want to be primary reviewer there.

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (ferantivero) [reply src/sign-in-xplat/MainPage.xaml.cs]: thanks for the feedback @ckittel, please let's confirm another level of inlining is desired in here. I understand the concept of having the code handy but at the same time I am worried about its readability.

partially done | addressed from [0c098](https://github.com/Azure-Samples/ms-identity-docs-code-csharp/pull/7/commits/0c098c645c61917802a67d423b30d3f2790c0708)

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (ferantivero) [reply src/sign-in-xplat/MainPage.xaml.cs]: I'm sure we can find some way of moving out the PublicClientApp `building` from here and make this more performant.

Please let's confirm we'd rather prefer performance over exposing/in-lining as a way to help the user see (without jumping context) what code they are responsible for.

Comment (ckittel) [reply src/sign-in-xplat/MainPage.xaml.cs]: I think we want to show proper lifecycle usage of the msal library.  In other languages we "new up the client" before we use it, external to where we use it.  Because that's the lifecycle of the client.  In the webapp version for python, you don't manage the client itself, but instead you need to manage the lifecycle of the token AND http cache, which you pass into the client.  Here, I'd imagine we'd just need to handle the lifecycle of the client.

If for some reason you feel it's best to leave it inline, then you'll want a clear comment above it indicating that this isn't demonstrating proper lifecycle management, I would think.

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (ckittel) [reply src/sign-in-xplat/MainPage.xaml.cs]: Good progress, like you said only partially, the next step is probably the most important, and that's getting the `SignInUserAndGetTokenUsingMSAL` bits inlined -- as that's the core of what we're teaching here, and with proper comments above every "step" along the way, it'll be very clear about what the logical flow of the process is.

Comment (ferantivero) [reply src/sign-in-xplat/MainPage.xaml.cs]: a bit more progress on the in-lining. We inlined the `PublicClientApp` instantiation. 

Please let me know if we want to inline the fill `SignInUserAndGetTokenUsingMSAL` method.

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (ferantivero) [reply src/sign-in-xplat/MainPage.xaml.cs]: thanks for the feedback @ckittel, we inlined it and we are now making a null check as a way to improve this.

done | addressed from [66539](https://github.com/Azure-Samples/ms-identity-docs-code-csharp/pull/11/commits/66539567652a2db7410b798a37792d8b828e4d84)

Comment (ckittel) [reply src/sign-in-xplat/MainPage.xaml.cs]: Lifecycle changes look good in #11, thanks.

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (ckittel) [reply src/sign-in-xplat/MainPage.xaml.cs]: Yeah, I think getting `SignInUserAndGetTokenUsingMSAL` brought into the function body here will be helpful for content development and fulfilling the "all in one place" request that Marsh has of the MSAL work being done.  After that, I'm seeing no "blockers" here for merge.

Comment (ferantivero) [reply src/sign-in-xplat/MainPage.xaml.cs]: sure thing @ckittel, all-in-lined now!


done | addressed from [`c294d`](https://github.com/Azure-Samples/ms-identity-docs-code-csharp/pull/11/commits/c294d7b3f6c0d0ee06f4a515d48236cb30029b17)

Comment (ckittel) [reply src/sign-in-xplat/MainPage.xaml.cs]: Thanks, looks like what we've been instructed to do. 💪

**Review by ckittel — APPROVED** [review]
Marking this Approved given that #11 is merged in here before merging to `main`.  w00t!

**Comment by ckittel** [comment]
@mmacy this is ready for Content Developer review of the README (plus anything else you'd like to review). Let @ferantivero know if there are any blockers before doing our initial merge. Thanks!

File: src/sign-in-xplat/README.md
Comment (mmacy) [inline src/sign-in-xplat/README.md:23]: Need output example.

File: src/sign-in-xplat/README.md
Comment (mmacy) [inline src/sign-in-xplat/README.md:68]: @ckittel We need a formatting convention for these replaceable token values - they're sorta all over the place in these samples. Then we need a pass through them all to adhere to the convention.

Unless the following goes against the language convention, I propose:

- short
- descriptive
- all CAPS
- NO_SPACES - separate words with underscore (`_`)

```csharp
// NOT our style:
private const string ClientId = "[Enter the Client Id (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]";

// Our style:
private const string _clientID = "APPLICATION_(CLIENT)_ID";
```

File: src/sign-in-xplat/README.md
Comment (mmacy) [inline src/sign-in-xplat/README.md:80]: These can be in the same section - there are only two steps.

Also, re:**Windows Machine**, I do see the that you say Visual Studio 2022 is a prereq - is this project really not supported in Visual Studio Code on a Linux machine? That seems nutty.

File: src/sign-in-xplat/README.md
Comment (mmacy) [inline src/sign-in-xplat/README.md:88]:
~~~suggestion
~~~

This only works if they've done the optional Azure CLI method of app registration. I recommend *not* including a *Clean Up* section.

**Review by mmacy — CHANGES_REQUESTED** [review]
Thanks @ferantivero - some changes required, and a question or two..

File: src/sign-in-xplat/README.md
Comment (mmacy) [inline src/sign-in-xplat/README.md:64]: Paths and filenames are italicized.

File: src/sign-in-xplat/README.md
Comment (ckittel) [reply src/sign-in-xplat/README.md:68]: I could totally see that working well.

This would be another great task for the content development team members who want to be involved in the creation of the code artifacts, like we heard.  Our team could pause and do this as well, but personally, I'd like to see our team continue to focus on the bulk of the scenario coverage and then leave any unification or presentational tweaks to those writing the presentation layer for this all.  Open to any path though here.

Comment (mmacy) [reply src/sign-in-xplat/README.md:68]: > I'd like to see our team continue to focus on the bulk of the scenario coverage and then leave any unification or presentational tweaks to those writing the presentation layer for this all.

I'm cool with that for the retroactive updates. Are you open to using that format for new samples, though?

Comment (ckittel) [reply src/sign-in-xplat/README.md:68]: Indeed!

File: src/sign-in-xplat/README.md
Comment (ferantivero) [reply src/sign-in-xplat/README.md:32]: thanks for feedback @mmacy 

done | addressed from [45d8d](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/45d8d3cdfc5f46a4d5cf48d6c174087d8f401f34) and [d00aa](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/d00aa944cb848e61bd570587bd2839de24646fd9)

File: src/sign-in-xplat/README.md
Comment (ferantivero) [reply src/sign-in-xplat/README.md:47]: good catch @mmacy 

done | address from [e58a8](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/e58a8bd8ee65aa6a204057a38f74d2ddc9f90c58)

File: src/sign-in-xplat/README.md
Comment (ferantivero) [reply src/sign-in-xplat/README.md:92]: thanks for feedback @mmacy 

done | attempting to address this [ca7a5](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/ca7a5919a3670dc72848c9097bef93824f379a0a)

File: src/sign-in-xplat/README.md
Comment (ferantivero) [reply src/sign-in-xplat/README.md:43]: yeah, it looks nicer @mmacy, but please just let me know in case you want to revert.


done | addressed from [fb8c4](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/fb8c4357be35f62bc212c782f9b15efdda3a6204)

File: src/sign-in-xplat/README.md
Comment (ferantivero) [reply src/sign-in-xplat/README.md:64]: thanks for pointing this out @mmacy 


done | addressed from [70bd2](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/70bd25a7c57661c77697ab34bac2ac61ba3ca07a)

File: src/sign-in-xplat/README.md
Comment (ferantivero) [reply src/sign-in-xplat/README.md:88]: that is absolutely @mmacy 

removed | addressed [cc93e](from https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/cc93ea488cfeaa78dc84722d4bf82405603b271c)

File: src/sign-in-xplat/README.md
Comment (ferantivero) [reply src/sign-in-xplat/README.md:80]: you are right @mmacy, please give another try.

bug fixed | addressed [9e5ba](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/9e5ba3b9a9c9152f0114c16563a74e6aea49f578)

File: src/sign-in-xplat/README.md
Comment (ferantivero) [reply src/sign-in-xplat/README.md:68]: thanks for the feedback @mmacy, for future changes on this we're looking forward to the retroactive updates. 

done | addressed from [3993d](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/3993d0cfda18a81c78cc8edc3ba5dc4b604d127b)

File: src/sign-in-xplat/README.md
Comment (ferantivero) [reply src/sign-in-xplat/README.md:23]: I got you @mmacy 

done | addressed from [bd39b](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/bd39b219b802630712d79645faa25bb239eaade2)

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (ferantivero) [reply src/sign-in-xplat/MainPage.xaml.cs:25]: hopefully this is only one missing convention @mmacy 

done | addressed from [c0337](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/c03379ea3f644a6556306ae964c8a92b269becd2)

File: src/sign-in-xplat/README.md
Comment (mmacy) [reply src/sign-in-xplat/README.md:92]: This gets the meat of it in there - app flow, libraries used, mentions consent. All good to have in there.

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (mmacy) [reply src/sign-in-xplat/MainPage.xaml.cs:25]: There's also `scopes`, but according to the linked conventions doc, all of these should be `_pascalCase` or `s_pascalCase` because they're private fields or private static fields.

Also, we should also prefer `readonly` over `const`, where possible (noted in point 3 in https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md).

File: src/sign-in-xplat/README.md
Comment (mmacy) [reply src/sign-in-xplat/README.md:64]: The https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/70bd25a7c57661c77697ab34bac2ac61ba3ca07a version is close:

`1. Open the _~/sign-in-xplat/MainPage.xaml.cs_ in your code editor.`

However, we shouldn't assume the project dir resides in the user's homedir (`~`). You could use this format or something like it that doesn't include the tilde (since we don't know where their repo clone or extracted archive resides):

- Open the _sign-in-xplat/MainPage.xaml.cs_ file in your editor.

File: src/sign-in-xplat/README.md
Comment (mmacy) [reply src/sign-in-xplat/README.md:80]: Looks like we have a split PR with #15, so I'll try when one of these is merged and the other is updated with its changes, or both are merged. If I wait until both are merged, I'll file a bug if it's still busted.

**Review by mmacy — CHANGES_REQUESTED** [review]
Getting closer, @ferantivero, but have some changes requested.

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (ferantivero) [reply src/sign-in-xplat/MainPage.xaml.cs:25]: thanks for detailing the conventions over here @mmacy 

bug fixing them take 2 | [29266](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/29266a01a94550f5268a1d075ebb15781700fd6c)

File: src/sign-in-xplat/README.md
Comment (ferantivero) [reply src/sign-in-xplat/README.md:64]: good call @mmacy 


done | [a35d5](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/15/commits/a35d5be30b798476f8c80887d9840ded74593791)

File: src/sign-in-xplat/MainPage.xaml.cs
Comment (mmacy) [reply src/sign-in-xplat/MainPage.xaml.cs:25]: Other than the PublicClientApplication, I'm not sure they all need to be static, but this is looking better - thanks, @ferantivero!

**Review by mmacy — APPROVED** [review]
LGTM, thanks @ferantivero. :shipit:

**Comment by ckittel** [comment]
Shipping!  :shipit:

**Comment by ferantivero** [comment]
w00t! thanks for the support on this @ckittel and @mmacy

**Pull request merged**
